File: src/trie/Trie.java
package trie;


/*
insert and search O(M), m is the length of the word
used in auto-completion, spell checking
 */
public class Trie {

    private final Node root;

    public Trie() {
        this.root = new Node('\0');
    }

    public void insert(String word) {
        Node curr = root;
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if (curr.children[c - 'a'] == null) curr.children[c - 'a'] = new Node(c);
            curr = curr.children[c - 'a'];
        }
        curr.isWord = true;
    }

    public boolean search(String word) {
        Node node = getNode(word);
        return node != null && node.isWord;
    }

    public boolean startsWith(String word) {
        Node node = getNode(word);
        return node != null;
    }

    private Node getNode(String word) {
        Node curr = root;
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);

            if (curr.children[c - 'a'] == null) return null;

            curr = curr.children[c - 'a'];

        }

        return curr;
    }

    static class Node {
        public char c;
        public boolean isWord;
        public Node[] children;

        public Node(char c) {
            this.c = c;
            this.isWord = false;
            this.children = new Node[26];
        }

    }
}

File: src/trie/kt/Trie.kt
package trie.kt

class Trie {

    private val root: Node = Node()

    fun insert(word: String) {
        var curr: Node? = root
        for (element in word) {
            if (curr?.children?.get(element - 'a') == null) curr?.children?.set(element - 'a', Node())
            if (curr != null) {
                curr = curr.children[element - 'a']
            }
        }
        curr?.isWord = true
    }

    private fun getNode(word: String): Node? {
        var curr: Node? = root
        for (element in word) {
            if (curr?.children?.get(element - 'a') == null) return null
            curr = curr.children[element - 'a']
        }
        return curr
    }

    fun search(word: String): Boolean {
        val node: Node? = getNode(word)
        return node != null && node.isWord
    }

    fun startsWith(word: String): Boolean {
        val node: Node? = getNode(word)
        return node != null
    }

    internal class Node(
        var isWord: Boolean = false,
    ) {
        var children: Array<Node?> = arrayOfNulls(26)
    }
}

fun main() {
    val trie = Trie()
    trie.insert("cat")
    trie.insert("cars")
    trie.insert("cushion")
    trie.insert("cooker")

    println(trie.search("cooker"))


    println(trie.startsWith("cars"))
}